cpp/src/arrow/compute/kernels/vector_sort_test.cc

```diff
@@ -1936,6 +1940,399 @@ TEST_P(TestTableSortIndicesRandom, Sort) {
   }
 }
 
+// ----------------------------------------------------------------------
+// Tests for Rank
+
+template <typename T>
+void AssertRank(const std::shared_ptr<T>& input, SortOrder order,
+                NullPlacement null_placement, RankOptions::Tiebreaker tiebreaker,
+                const std::shared_ptr<Array>& expected) {
+  RankOptions options(order, null_placement, tiebreaker);
+  ASSERT_OK_AND_ASSIGN(auto actual, CallFunction("rank", {input}, &options));
+  AssertDatumsEqual(expected, actual, /*verbose=*/true);
+}
+
+void AssertRank(const std::shared_ptr<DataType>& type, const std::string& values,
+                SortOrder order, NullPlacement null_placement,
+                RankOptions::Tiebreaker tiebreaker, const std::string& expected) {
+  AssertRank(ArrayFromJSON(type, values), order, null_placement, tiebreaker,
+             ArrayFromJSON(uint64(), expected));
+}
+
+class TestRankBase : public ::testing::Test {
+ public:
+  virtual std::shared_ptr<DataType> type() = 0;
+
+  virtual void AssertRank(const std::string& values, SortOrder order,
+                          NullPlacement null_placement,
+                          RankOptions::Tiebreaker tiebreaker,
+                          const std::string& expected) {
+    arrow::compute::AssertRank(this->type(), values, order, null_placement, tiebreaker,
+                               expected);
+  }
+};
+
+template <typename ArrowType>
+class TestRank : public TestRankBase {
+ public:
+  std::shared_ptr<DataType> type() override {
+    // Will choose default parameters for temporal types
+    return std::make_shared<ArrowType>();
+  }
+};
+
+template <typename ArrowType>
+class TestRankForReal : public TestRank<ArrowType> {};
+TYPED_TEST_SUITE(TestRankForReal, RealArrowTypes);
+
+template <typename ArrowType>
+class TestRankForBool : public TestRank<ArrowType> {};
+TYPED_TEST_SUITE(TestRankForBool, ::testing::Types<BooleanType>);
+
+template <typename ArrowType>
+class TestRankForIntegral : public TestRank<ArrowType> {};
+TYPED_TEST_SUITE(TestRankForIntegral, IntegralArrowTypes);
+
+template <typename ArrowType>
+class TestRankForTemporal : public TestRank<ArrowType> {};
+TYPED_TEST_SUITE(TestRankForTemporal, TemporalArrowTypes);
+
+template <typename ArrowType>
+class TestRankForStrings : public TestRank<ArrowType> {};
+TYPED_TEST_SUITE(TestRankForStrings, StringSortTestTypes);
+
+class TestRankForFixedSizeBinary : public TestRankBase {
+ public:
+  std::shared_ptr<DataType> type() override { return fixed_size_binary(3); }
+};
+
+TYPED_TEST(TestRankForReal, RankReal) {
```

Comment:
I know we use type-parameterized tests a lot, but we should refrain from adding too many of them especially when the actual test code doesn't vary (the type parameter appears nowhere below AFAICT), because the compilation time for tests is not trivial.

The nice thing with JSON-specified test values is that the same code can be run just with a different `shared_ptr<DataType>` argument (and possibly a different JSON string, if we want to get sophisticated).

So there are two possible alternatives:
* use value-parameterized tests, with the various `shared_ptr<DataType>` instances as parameter
* use a simple for loop on supported type instances